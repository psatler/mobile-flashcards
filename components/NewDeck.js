import React, { Component } from 'react'
import { Text, StyleSheet, TextInput, 
    Platform, TouchableOpacity, KeyboardAvoidingView, Dimensions } from 'react-native'
import { lightBlue, darkBlue } from '../utils/colors';

import { saveDeckTitle } from '../utils/asyncDB'
import { connect } from 'react-redux'
import { addDeck } from '../actions'

import { ImagePicker, Permissions } from 'expo'

export class NewDeck extends Component { //exporting for testing purposes

    //local state for handling form inputs
    state = {
        input: '',
        image: null, //to store image URI if the user uploads an image
    }

    handleTextChange = (deckTitle) => {
        this.setState({
            input: deckTitle,
        })
    }

    submitDeck = () => {
        const { input, image } = this.state;

        if(input){
            const deck = {
                title: input,
                image: image,
                questions: [],
                createdAt: new Date(),
            }

            const keyTitle = input.split(' ').join(''); //taking out the spaces to store key without'em

            //update redux
            this.props.dispatch(addDeck({ [keyTitle]: deck }));

            //reset state
            this.setState({
                input: '',
                image: null,
            })

            this.props.navigation.navigate('DeckDetail', { //go to details screen of the newly created deck
                deckTitle: deck.title,
                deckLength: deck.questions.length
            })

            //save to DB (async storage)
            saveDeckTitle(keyTitle, deck);
        }
    }

    askPermission = async () => {
        let res = await Permissions.askAsync(Permissions.CAMERA_ROLL)
        if(res.status === 'granted'){
            return this.pickImage()
        }
    }

    pickImage = async () => {
        let result = await ImagePicker.launchImageLibraryAsync({ //wait till it resolves
            allowsEditing: true,
            aspect: [4, 3],
        });

        if(result.cancelled){
            return
        }
        else {
            this.setState({ image: result.uri });
        }
    }

    render() {
        const { input } = this.state;

        //using keyboardAvoidingView to avoid keyboard hiding other views (buttons, textInputs, etc)
        return (
            <KeyboardAvoidingView style={styles.container} behavior='padding'>
                <Text style={styles.headline}> What's is the title of your new deck? </Text>
                <TextInput 
                    value={input}
                    placeholder={'Deck Title'}
                    style={styles.textInputStyle}
                    onChangeText={this.handleTextChange}
                />

                <TouchableOpacity style={styles.submitButton} onPress={this.askPermission}>
                    <Text style={styles.submitButtonText} > Open Camera Roll </Text>
                </TouchableOpacity>

                <TouchableOpacity style={styles.submitButton} onPress={this.submitDeck}>
                    <Text style={styles.submitButtonText} > Submit </Text>
                </TouchableOpacity>
            </KeyboardAvoidingView>
        )
    }
}

const window = Dimensions.get('window');
const styles = StyleSheet.create({
    container: {
      flex: 1,
      padding: 10,
      alignItems: 'center',
      justifyContent: 'center',
    },
    headline: {
        textAlign: 'center',
        fontSize: 40,
        color: lightBlue,
    },
    textInputStyle: {
        paddingLeft: 15,
        marginTop: 30,
        height: 50,
        borderColor: lightBlue,
        borderWidth: 1,
        borderRadius: Platform.OS === 'ios' ? 16 : 2,
        width: window.width - 30,
    },
    submitButton: {
        marginTop: 25,
        padding: 10,
        borderRadius: Platform.OS === 'ios' ? 16 : 2,
        backgroundColor: darkBlue,
    },
    submitButtonText: {
        fontSize: 15,
        color: lightBlue,
        padding: 5,
    }
});

export default connect()(NewDeck);



