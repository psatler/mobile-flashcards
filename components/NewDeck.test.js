import React from 'react'
import { NewDeck } from './NewDeck'
import renderer from 'react-test-renderer';
// import ShallowRenderer from 'react-test-renderer/shallow';

//###### ENZYME STUFF
import { configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
configure({ adapter: new Adapter() })
import { shallow } from 'enzyme'

describe('[Component] New Deck', () => {

    it('shallow render correctly', () => {
        const wrapper = shallow(<NewDeck />);
        expect(wrapper).toMatchSnapshot();
    });

    it('should press the submit button', () => {
        const wrapper = shallow(<NewDeck />);
        wrapper.instance().submitDeck();
        expect(wrapper.state("input")).toEqual('');
        expect(wrapper.state("image")).toEqual(null);

        //############ TRYING TO CREATE A SPY - NO SUCCESS ##########
        //here https://github.com/airbnb/enzyme/issues/365 it says I should've used constructor instead of arrow functions
        // const spy = jest.spyOn(NewDeck.prototype,'submitDeck');
        // const wrapper = shallow(<NewDeck />);
        // wrapper.find('TouchableOpacity').last().props().onPress()
        // expect(spy).toHaveBeenCalled();
        // console.log(NewDeck.prototype)
    });

    it('should call askPermission method', () => {
        const wrapper = shallow(<NewDeck />);
        wrapper.instance().askPermission();
    })
    
});