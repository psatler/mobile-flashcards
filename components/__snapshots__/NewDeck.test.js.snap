// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component] New Deck shallow render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NewDeck />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "behavior": "padding",
      "children": Array [
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#00A4D3",
              "fontSize": 40,
              "textAlign": "center",
            }
          }
        >
           What's is the title of your new deck? 
        </Text>,
        <TextInput
          allowFontScaling={true}
          onChangeText={[Function]}
          placeholder="Deck Title"
          style={
            Object {
              "borderColor": "#00A4D3",
              "borderRadius": 16,
              "borderWidth": 1,
              "height": 50,
              "marginTop": 30,
              "paddingLeft": 15,
              "width": 720,
            }
          }
          value=""
        />,
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
          style={
            Object {
              "backgroundColor": "#18263D",
              "borderRadius": 16,
              "marginTop": 25,
              "padding": 10,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#00A4D3",
                "fontSize": 15,
                "padding": 5,
              }
            }
          >
             Open Camera Roll 
          </Text>
        </TouchableOpacity>,
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
          style={
            Object {
              "backgroundColor": "#18263D",
              "borderRadius": 16,
              "marginTop": 25,
              "padding": 10,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#00A4D3",
                "fontSize": 15,
                "padding": 5,
              }
            }
          >
             Submit 
          </Text>
        </TouchableOpacity>,
      ],
      "enabled": true,
      "keyboardVerticalOffset": 0,
      "style": Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "padding": 10,
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "accessible": true,
          "allowFontScaling": true,
          "children": " What's is the title of your new deck? ",
          "ellipsizeMode": "tail",
          "style": Object {
            "color": "#00A4D3",
            "fontSize": 40,
            "textAlign": "center",
          },
        },
        "ref": null,
        "rendered": " What's is the title of your new deck? ",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "allowFontScaling": true,
          "onChangeText": [Function],
          "placeholder": "Deck Title",
          "style": Object {
            "borderColor": "#00A4D3",
            "borderRadius": 16,
            "borderWidth": 1,
            "height": 50,
            "marginTop": 30,
            "paddingLeft": 15,
            "width": 720,
          },
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "activeOpacity": 0.2,
          "children": <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#00A4D3",
                "fontSize": 15,
                "padding": 5,
              }
            }
          >
             Open Camera Roll 
          </Text>,
          "onPress": [Function],
          "style": Object {
            "backgroundColor": "#18263D",
            "borderRadius": 16,
            "marginTop": 25,
            "padding": 10,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "accessible": true,
            "allowFontScaling": true,
            "children": " Open Camera Roll ",
            "ellipsizeMode": "tail",
            "style": Object {
              "color": "#00A4D3",
              "fontSize": 15,
              "padding": 5,
            },
          },
          "ref": null,
          "rendered": " Open Camera Roll ",
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "activeOpacity": 0.2,
          "children": <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#00A4D3",
                "fontSize": 15,
                "padding": 5,
              }
            }
          >
             Submit 
          </Text>,
          "onPress": [Function],
          "style": Object {
            "backgroundColor": "#18263D",
            "borderRadius": 16,
            "marginTop": 25,
            "padding": 10,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "accessible": true,
            "allowFontScaling": true,
            "children": " Submit ",
            "ellipsizeMode": "tail",
            "style": Object {
              "color": "#00A4D3",
              "fontSize": 15,
              "padding": 5,
            },
          },
          "ref": null,
          "rendered": " Submit ",
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "behavior": "padding",
        "children": Array [
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#00A4D3",
                "fontSize": 40,
                "textAlign": "center",
              }
            }
          >
             What's is the title of your new deck? 
          </Text>,
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="Deck Title"
            style={
              Object {
                "borderColor": "#00A4D3",
                "borderRadius": 16,
                "borderWidth": 1,
                "height": 50,
                "marginTop": 30,
                "paddingLeft": 15,
                "width": 720,
              }
            }
            value=""
          />,
          <TouchableOpacity
            activeOpacity={0.2}
            onPress={[Function]}
            style={
              Object {
                "backgroundColor": "#18263D",
                "borderRadius": 16,
                "marginTop": 25,
                "padding": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#00A4D3",
                  "fontSize": 15,
                  "padding": 5,
                }
              }
            >
               Open Camera Roll 
            </Text>
          </TouchableOpacity>,
          <TouchableOpacity
            activeOpacity={0.2}
            onPress={[Function]}
            style={
              Object {
                "backgroundColor": "#18263D",
                "borderRadius": 16,
                "marginTop": 25,
                "padding": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#00A4D3",
                  "fontSize": 15,
                  "padding": 5,
                }
              }
            >
               Submit 
            </Text>
          </TouchableOpacity>,
        ],
        "enabled": true,
        "keyboardVerticalOffset": 0,
        "style": Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "padding": 10,
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "accessible": true,
            "allowFontScaling": true,
            "children": " What's is the title of your new deck? ",
            "ellipsizeMode": "tail",
            "style": Object {
              "color": "#00A4D3",
              "fontSize": 40,
              "textAlign": "center",
            },
          },
          "ref": null,
          "rendered": " What's is the title of your new deck? ",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "allowFontScaling": true,
            "onChangeText": [Function],
            "placeholder": "Deck Title",
            "style": Object {
              "borderColor": "#00A4D3",
              "borderRadius": 16,
              "borderWidth": 1,
              "height": 50,
              "marginTop": 30,
              "paddingLeft": 15,
              "width": 720,
            },
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "activeOpacity": 0.2,
            "children": <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#00A4D3",
                  "fontSize": 15,
                  "padding": 5,
                }
              }
            >
               Open Camera Roll 
            </Text>,
            "onPress": [Function],
            "style": Object {
              "backgroundColor": "#18263D",
              "borderRadius": 16,
              "marginTop": 25,
              "padding": 10,
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "accessible": true,
              "allowFontScaling": true,
              "children": " Open Camera Roll ",
              "ellipsizeMode": "tail",
              "style": Object {
                "color": "#00A4D3",
                "fontSize": 15,
                "padding": 5,
              },
            },
            "ref": null,
            "rendered": " Open Camera Roll ",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "activeOpacity": 0.2,
            "children": <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#00A4D3",
                  "fontSize": 15,
                  "padding": 5,
                }
              }
            >
               Submit 
            </Text>,
            "onPress": [Function],
            "style": Object {
              "backgroundColor": "#18263D",
              "borderRadius": 16,
              "marginTop": 25,
              "padding": 10,
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "accessible": true,
              "allowFontScaling": true,
              "children": " Submit ",
              "ellipsizeMode": "tail",
              "style": Object {
                "color": "#00A4D3",
                "fontSize": 15,
                "padding": 5,
              },
            },
            "ref": null,
            "rendered": " Submit ",
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
